### PR TITLE
[Wallet] Clean wallet auto-backup "feature"

### PR DESCRIPTION
**Pull request by furszy**
Was checking #2362 and wasn't able to contain myself from cleaning redundancies and code deduplication on the wallet auto-backup "feature".

### CODE REVIEW

**Comment by furszy** [comment]
Now that #2362 was merged, this is ready to go.

File: src/wallet/walletdb.cpp [hunk 1]
```diff
@@ -965,59 +913,66 @@ bool AutoBackupWallet(const std::string& strWalletFile, std::string& strBackupWa
...
+        LogPrintf("%s\n", strBackupWarning);
+        return false;
+    }
+
+    // Try to backup
+    if (!AttemptBackupWallet(nullptr, sourceFile, backupFile)) {
+        return false; // error is logged internally
+    }
+
+    // Keep only the last 10 backups, including the new one of course
```
Comment (random-zebra) [inline src/wallet/walletdb.cpp:952]: To be precise we keep `0 < nWalletBackups <= 10` backups.

**Review by random-zebra — CHANGES_REQUESTED** [review]
Concept ACK.
Although, as the tests show, this has a bug. See comments inlined.

Aside from that, it would be good to rebase this one on top of #2351 (which is already based on #2337) to minimize conflicts.

Also, basing this work on top of #2351, you could do another step in the refactoring.
At the moment we have an ordering issue: on the first run, `InitAutoBackupWallet` is called before creating the actual wallet (which is done in step 8, `InitLoadWallet`), so it always logs a warning and it never backups the file.
Would be good to create the new wallet, and *then* back it up.

In #2351 we move the backup step inside `InitLoadWallet`, so, with a little more refactoring, we could backup the actual __running/loaded__ wallets, calling directly `BackupWallet` (merging it with `AutoBackupWallet`, essentially flushing the database before copying the file).

**Comment by furszy** [comment]
Added a bit more work here.

1) Backup creation after the wallet initialization.
2) Removed the inexistent "-printcoinstake" init flag from the help command.
3) Added seconds to backup filename so it doesn't conflict with backups created closing-opening the wallet under the same minute.

**Comment by furszy** [comment]
Rebased on master, fixing conflicts. Ready to go.

File: src/wallet/init.cpp [hunk 2]
```diff
@@ -221,11 +220,17 @@ bool InitLoadWallet()
...
+        bool fFirstRun;
+        CWallet * const pwallet = CWallet::CreateWalletFromFile(walletFile, fFirstRun);
         if (!pwallet) {
             return false;
         }
         vpwallets.emplace_back(pwallet);
+
+        if (fFirstRun) {
+            // Initial backup
+            return AutoBackupWallet(walletFile, strWarning, strError);
```
Comment (random-zebra) [inline src/wallet/init.cpp:232]: This makes the multiwallet test fail. We are inside a loop `for (const std::string& walletFile : gArgs.GetArgs("-wallet"))`.
If we `return`, then only the first wallet is processed.

Comment (random-zebra) [reply src/wallet/init.cpp:232]: Also we ignore `strWarning`/`strError` if set.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -4346,7 +4346,7 @@ bool CWalletTx::AcceptToMemoryPool(CValidationState& state, bool fLimitFree, boo
 
 std::string CWallet::GetUniqueWalletBackupName() const
 {
-    return strprintf("%s%s", (dbw ? dbw->GetName() : "null"), FormatISO8601DateTimeForBackup(GetTime()));
+    return strprintf("%s%s", (dbw ? dbw->GetName() : "null"), FormatISO8601DateTime(GetTime()));
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/wallet.cpp:4349]: k, missed that. will just remove the warning and only log it.

**Comment by furszy** [comment]
> Wouldn't it be cleaner to just load/create the wallet before trying the backup, and keep the warning for missing source file 

We spoke about this few times already, the backup at startup was added in case of having a problem at the initialization & load phase and for some reason the wallet file gets corrupted, so the file gets copied **before** loading it. Different software versions could try to open the wallet db in different manners, even modify it (we should try to avoid anything like this..), so having a simple file copy backup before doing any operation over the file is by far simpler than parse and load it and then re-serialize everything to save it to disk again.

> Currently, inside InitLoadWallet, we are doing this :
1. try to backup. if the wallet file is not found, don't warn.
2. try to load the wallet. if it doesn't exist, create a new one, and set fFirstRun to true.
3. if fFirstRun is true, then try the backup once again.

Not sure if if you are thinking that it will do the backup twice with the last "once again". In case you are thinking it, it will not happen. 
If the file doesn't exist (point 1 does nothing aka file not found) then point 3 makes the first backup after the wallet creation. If point 1 does the backup (file exist, wallet already created), then point 3 does nothing (not in the first run, `fFirstRun=false`).

File: src/wallet/init.cpp (hunk 2 above)
Comment (furszy) [reply src/wallet/init.cpp:232]: yeah, will add a log here instead of the return and done. No need to overthink this much.

**Comment by random-zebra** [comment]
> Not sure if you are thinking that it will do the backup twice

It will *attempt* to do it twice. First time it will fail of course. That's why you had to remove the warning. 
And the second time it is done *after* initialization.
Anyway, no big deal... provided that we don't immediately return, breaking multiwallet initialization, and ignoring warnings/errors.

**Comment by furszy** [comment]
You know something, thinking this further, aside from the refactoring and code cleanup that made here, the problem that have with this flow is the following one: if the wallet file is already corrupted and the user tries to open the wallet +10 times, then every backup file will be a copy of the corrupted file, the rolling backup process will remove the good backups and create only corrupted ones.

Will move it to the backup after the unserialization and wallet load, it could introduce corruptions only if we do nasty nasty things on the init/load process such as a data write etc (which will not happen while we are around..)

**Comment by furszy** [comment]
Updated now, cherry-picked your commit.

**Review by random-zebra — APPROVED** [review]
ACK f36c4d04b20aed3cbcbedeed210858bcc3649dd4

**Review by Fuzzbawls — APPROVED** [review]
ACK ce4ba966f25d7c82c5e2484a86ee5efc102704f2

**Pull request merged**